docs/node/grpc.md

```diff
@@ -0,0 +1,285 @@
+- authentication,
+- load balances the traffic to multiple Oasis nodes.
+
+This chapter will show you how to set up a **public** gRPC endpoint using Envoy
+so that it exposes only a **safe subset of the [Oasis RPC services]**. The
+final section presents an alternative approach which **securely tunnels a Unix
+socket over the network**, so it can safely be used by the client, but **does
+not filter out any services**.
+
+:::danger Never expose RPC Unix socket directly!
+
+The `oasis-node` deliberately exposeses the RPC interface only via an AF_LOCAL
+socket called `internal.sock` located in the node's data directory.
+The socket should **never be directly exposed over the network** as it has no
+authentication and allows full control—including shutting down your node.
+
+:::
+
+[gRPC interface]: https://github.com/oasisprotocol/oasis-core/blob/master/docs/oasis-node/rpc.md
+[Web3 gateway]: web3.md
+[Oasis RPC services]: https://github.com/oasisprotocol/oasis-core/blob/master/docs/oasis-node/rpc.md#services
+
+## gRPC Proxy with Envoy {#envoy}
+
+Let's set up a typical public Oasis endpoint using the [Envoy HTTP proxy] with
+the following behavior:
+
+- whitelisted methods for checking the network status, performing queries and
+  submitting transactions and no control methods and queries that are CPU and
+  I/O intensive,
+- lives on `grpc.my-oasis-node.com` with TLS-enabled connection and Let's
+  encrypt certificates.
+- `internal.sock` of the Oasis node accessible at `/node/data/internal.sock`.
+
+Envoy already has built-in support for gRPC so after installing Envoy, simply
+put the configuration below inside your `envoy.yaml` and then execute the proxy
+with `envoy -c envoy.yaml`.
+
+```yaml title="envoy.yaml"
+# Envoy gRPC-web proxy configuration
+---
+admin:
+  address:
+    socket_address:
+      address: "127.0.0.1"
+      port_value: 10000
+static_resources:
+  listeners:
+    - name: listener_0
+      address:
+        socket_address:
+          address: "0.0.0.0"
+          port_value: 443
+      filter_chains:
+        - filters:
+            - name: envoy.filters.network.http_connection_manager
+              typed_config:
+                "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
+                codec_type: AUTO
+                stat_prefix: ingress_http
+                access_log:
+                  - name: envoy.file_access_log
+                    typed_config:
+                      "@type": type.googleapis.com/envoy.extensions.access_loggers.file.v3.FileAccessLog
+                      path: /dev/stdout
+                route_config:
+                  virtual_hosts:
+                    - name: vh_0
+                      domains:
+                        - "*"
+                      routes:
+                        - match:
+                            safe_regex:
+                              regex: "/oasis-core\\.(\
+    Beacon/(\
```

Comment:
oh this might not work when it's spaced out like this. we have it in this format in our internal repo, but with an additional filter later to remove spaces

Reply:
Using `"` will trim leading whitespaces of every line and it will replace the new lines with a single whitespace (which we do not want). Then, using `\` at the end of each new line will make the newlines not to be replaced with a whitespace.

Ref: https://yaml.org/spec/1.2.2/#731-double-quoted-style

Tested it out here: https://eemeli.org/yaml-playground/

Reply:
oo nice find